### PR TITLE
feat(core): verifyAndParseCodeFromCallbackUri

### PR DESCRIPTION
**Pull request by xiaoyijun**
## Summary

* feat(core): verifyAndParseCodeFromCallbackUri

## Linear Issue Reference

* LOG-746
* LOG-989

## Testing

image

@logto-io/eng

### CODE REVIEW

File: kotlin/src/main/kotlin/io/logto/sdk/core/constant/QueryKey.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package io.logto.sdk.core.constant
+
+object QueryKey {
```
Comment (simeng-li) [inline kotlin/src/main/kotlin/io/logto/sdk/core/constant/QueryKey.kt:3]: Feel like a little bit too general. Especially under a shared constant file directly under the core.

File: kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+open class LogtoException(
+    message: String,
+    cause: Throwable? = null,
+) : RuntimeException(message, cause) {
+    class VerificationException(
+        message: Enum<Verification>,
+        cause: Throwable? = null,
+    ) : LogtoException(message.name, cause)
+
+    enum class Verification {
```
Comment (simeng-li) [inline kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt:12]: Similarly here, should it be sth like CallbackUrlVerification? Verification seems should also cover i.e idToken verification errors.

File: kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        INVALID_URI_FORMAT,
+        URI_MISMATCHED,
+        MISSING_CODE_URI_PARAMETER,
+        MISSING_STATE_URI_PARAMETER,
+        STATE_MISMATCHED,
+    }
+
+    class RedirectUriReturnedException(
```
Comment (simeng-li) [inline kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt:20]: Do we need an extra class explicitly for this case? Or just map the error and errorDesc values to the general verificationException you have defined above.

File: kotlin/src/main/kotlin/io/logto/sdk/core/util/UriUtils.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    fun verifyAndParseCodeFromCallbackUri(
+        callbackUri: String,
+        redirectUri: String,
+        state: String,
+    ): String {
+        if (!callbackUri.startsWith(redirectUri)) {
+            throw LogtoException.VerificationException(LogtoException.Verification.URI_MISMATCHED)
+        }
+
+        val parsedUri = callbackUri.toHttpUrlOrNull()
```
Comment (simeng-li) [inline kotlin/src/main/kotlin/io/logto/sdk/core/util/UriUtils.kt:17]: not sure, but, if we have passed `callbackURI` start with `redirectUri` validation above, could it still be an invalid Uri here?

File: kotlin/src/test/kotlin/io/logto/sdk/core/util/UriUtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        val expectedException = Assert.assertThrows(LogtoException.RedirectUriReturnedException::class.java) {
+            UriUtils.verifyAndParseCodeFromCallbackUri(callbackUri, redirectUri, state)
+        }
+
+        assertThat(expectedException.error).isEqualTo(error)
+        assertThat(expectedException.errorDesc).isNull()
+    }
+
+    @Test
+    fun verifyAndParseCodeFromCallbackUriShouldThrowWithBothErrorDescAndErrorParameter() {
```
Comment (simeng-li) [inline kotlin/src/test/kotlin/io/logto/sdk/core/util/UriUtilsTest.kt:74]: maybe rename to should show errorDesc message if present in the callback URL. As errorDesc itself takes no effect on throwing an exception right. i.e. if no error is present, just an errorDesc there won't be any exceptions at all. This case is more like to test the exception should contains errorDesc content.

File: kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package io.logto.sdk.core.exception
+
+open class LogtoException(
+    message: String,
+    cause: Throwable? = null,
+) : RuntimeException(message, cause) {
+    class VerificationException(
+        message: Enum<Verification>,
+        cause: Throwable? = null,
+    ) : LogtoException(message.name, cause)
```
Comment (gao-sun) [inline kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt:10]: 为啥这里套娃了？能不能直接继承 `RuntimeException`

File: kotlin/src/main/kotlin/io/logto/sdk/core/constant/QueryKey.kt (hunk 1 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/constant/QueryKey.kt:3]: I'm ok with it.
We will have `QueryKey`, `GrantType`, and other constants here. All of them are used in the core.

File: kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt (hunk 2 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt]: updated

File: kotlin/src/main/kotlin/io/logto/sdk/core/util/UriUtils.kt (hunk 4 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/util/UriUtils.kt:19]: refer to the [slack thread](https://silverhand-io.slack.com/archives/C02A8G4HVAM/p1642580813321000)

File: kotlin/src/test/kotlin/io/logto/sdk/core/util/UriUtilsTest.kt (hunk 5 above)
Comment (xiaoyijun) [reply kotlin/src/test/kotlin/io/logto/sdk/core/util/UriUtilsTest.kt:76]: updated

File: kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt (hunk 6 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt:10]: 直接继承 `RuntimeException`，后续要捕获这些错误就得捕获 `RuntimeException`，捕获 `RuntimeException` 不被 lint 允许，因为这可能吞掉非 Logto 产生的异常

统一放在 `LogtoException` 底下，是计划后面可以通过捕获 `LogtoException` 来捕获我们 SDK 给出的异常类型

File: kotlin/src/main/kotlin/io/logto/sdk/core/util/UriUtils.kt (hunk 4 above)
Comment (xiaoyijun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/util/UriUtils.kt]: updated

**Review by gao-sun — APPROVED** [review]
lgtm

File: kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt (hunk 6 above)
Comment (gao-sun) [reply kotlin/src/main/kotlin/io/logto/sdk/core/exception/LogtoException.kt]: discussed offline

**Review by simeng-li — APPROVED** [review]
LGTM

**Pull request merged**